Scheduler metrics: binding rate limiter saturation rate

hongchaodeng opened the pull request:
This PR exposes the pod binding rate. We are exposing it because there is a rate limiter. We want to measure and see how long before it becomes a bottleneck.

xiang90 commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -107,6 +107,13 @@ func New(c *Config) *Scheduler {
 
 // Run begins watching and scheduling. It starts a goroutine and returns immediately.
 func (s *Scheduler) Run() {
+	go func() {
```
can we move this to metrics.go and add a func called `ReportRateLoop` or something like that?

hongchaodeng commented:
Made some wrong assumptions on rate limiter.
Working on better solutions now.

hongchaodeng commented:
Once https://github.com/juju/ratelimit/pull/10 is merged, I should be able to continue on this work:
- [ ] bump up juju/ratelimit version
- [ ] expose Available() and Capacity() and use it to export rate limiter's saturation metrics

davidopp commented:
@wojtek-t

davidopp commented:
@gmarek

wojtek-t commented on plugin/pkg/scheduler/metrics/metrics.go:
```diff
@@ -20,11 +20,15 @@ import (
 
 const schedulerSubsystem = "scheduler"
 
+var BindingRateObserveInterval = 200 * time.Millisecond
```
make it const

wojtek-t commented on plugin/pkg/scheduler/metrics/metrics.go:
```diff
@@ -20,11 +20,15 @@ import (
 const schedulerSubsystem = "scheduler"
 
+var BindingRateObserveInterval = 200 * time.Millisecond
+
```
Also - do you think that exposing it every 200ms is necessary? I wouldn't go below 1s...

wojtek-t commented:
@hongchaodeng - what's the status of it? Do you want to proceed with it?

hongchaodeng commented:
@wojtek-t 
I'm currently using it along with other metrics to meaningful insights.
Let's just close this.
I will send it in another PR once I have more complete story.
Thanks!

xiang90 commented:
@hongchaodeng I think we should push in the rate limiter saturation metrics anyway.

wojtek-t commented:
I agree with @xiang90 

@hongchaodeng - can you please rebase this PR and answer my comments so that we can proceed with that?

hongchaodeng committed:
godep: bump up juju/ratelimit

hongchaodeng committed:
change tickRateLimiter to tokenBucketRateLimiter

hongchaodeng commented:
@wojtek-t 

- Rewrote `Rate()` to `CanAccept()`. The reason is that `Rate()` is not quite generic for all rate limiters but `CanAccept()` is. We still export saturation status of rate limiter to be gauge metrics of 0/1.
- Added test for `CanAccept()`. We can't test metrics so I manually set up local cluster and verified  it.
- Godep: bump up `juju/ratelimit` version to make use of the newly added interface `Available()` which shows current available tokens.

PTAL. Thanks!

hongchaodeng commented:
This issue fixed #17762. 
@gmarek

lavalamp commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -107,6 +107,7 @@ func New(c *Config) *Scheduler {
 
 // Run begins watching and scheduling. It starts a goroutine and returns immediately.
 func (s *Scheduler) Run() {
+	go metrics.ReportBindingRateLimiterSaturation(s.config.BindPodsRateLimiter)
```
Perhaps add a name to rate-limiter creation, and have rate limiters auto-register themselves?

hongchaodeng replied:
This is specific to binding rate limiter.
The metrics/ dir in also under scheduler/ .
If you want a generic rate limiter, I doubt that's gonna be large changes, which is beyond initial attempt for scalability testing purpose.
Can we raise another issue and ask metrics related folks to do it?

lavalamp replied:
OK, let's handle auto registration separately.

lavalamp commented on plugin/pkg/scheduler/metrics/metrics.go:
```diff
@@ -68,3 +80,17 @@ func Register() {
+
+// ReportBindingRateLimiterSaturation exports status of binding rate limiter
+// whether it's full (1) or not (0).
+func ReportBindingRateLimiterSaturation(rl util.RateLimiter) {
```
This is not binding specific code, so remove that from the name (and pass the metric to set in as a parameter).

hongchaodeng replied:
Right now, it is a binding specific. But, yes, we can change the function to be more generic. Like ReportRateLimiterSatruation(typ string, saturated bool)?

lavalamp commented on plugin/pkg/scheduler/metrics/metrics.go:
```diff
@@ -68,3 +80,17 @@ func Register() {
+func ReportBindingRateLimiterSaturation(rl util.RateLimiter) {
+	for {
+		if rl.CanAccept() {
+			BindingRateLimiterSturation.Set(0)
```
Seems like you can do a lot better by using 1 - (# tokens available) / (# tokens max). Then you could perhaps reduce the rate.

The way you have it now, suppose it's *always* used 19 out of 20 tokens. Then your metric will show 0, when really it should be .95. If code changes ever so slightly to use one more, it will instantly switch from 0 to 1.

hongchaodeng replied:
That's correct and also my original thoughts.
But see my comments:
> Rewrote Rate() to CanAccept(). The reason is that Rate() is not quite generic for all rate limiters but CanAccept() is. We still export saturation status of rate limiter to be gauge metrics of 0/1.

wojtek-t replied:
I don't have strong opinion on the api, but we definitely should export the saturation, not just 0 or 1 (which isn't very informative).

hongchaodeng committed:
Rate limiter: Add Saturation() method

timothysc commented on plugin/pkg/scheduler/metrics/metrics.go:
```diff
@@ -61,6 +72,7 @@ func Register() {
 		prometheus.MustRegister(E2eSchedulingLatency)
 		prometheus.MustRegister(SchedulingAlgorithmLatency)
 		prometheus.MustRegister(BindingLatency)
+		prometheus.MustRegister(BindingRateLimiterSturation)
```
s/Sturation/Saturation

hongchaodeng replied:
Good catch! Thanks.

timothysc commented on plugin/pkg/scheduler/metrics/metrics.go:
```diff
@@ -50,6 +54,13 @@ var (
 			MaxAge:    time.Hour,
 		},
 	)
+	BindingRateLimiterSturation = prometheus.NewGauge(
```
spell check on Saturation

hongchaodeng commented:
Great! Thanks @lavalamp and @wojtek-t 's comments. I have moved forward and exposed the saturation rate in percentage.

Re. Daniel's thoughts on auto-register, I don't have any idea how to do it and do not think it should be done here either. Can we create another issue? I think it makes more sense since it's another functionality.

Thanks again!

timothysc commented on pkg/util/throttle.go:
```diff
@@ -22,13 +22,16 @@ type RateLimiter interface {
 }
 
-type tickRateLimiter struct {
+type tokenBucketRateLimiter struct {
```
the name change seems orthogonal to the change of the metric, but Meh.

lavalamp commented on pkg/util/throttle.go:
```diff
@@ -26,9 +26,12 @@ type RateLimiter interface {
 	// Stop stops the rate limiter, subsequent calls to CanAccept will return false
 	Stop()
+	// Saturation returns a percentage number which describes how saturated
+	// this rate limiter is.
```
Add: "1.0 means no tokens are available, 0.0 means all tokens are available."

gmarek replied:
I'd suggest a slightly different semantics: there are 2 different 'full' levels, all burst tokens are gone, and we're using tokens in the same rate we're producing them (i.e. we're at full capacity of 'normal' rate). I'd rather have 1.0 mean the latter, and values above 1.0 the former (concrete value depends on how much given RL can burst)

hongchaodeng replied:
@gmarek That doesn't sound like Saturation() but Available() rate.

lavalamp replied:
@gmarek I don't really follow how that would make a difference. There's no difference between "burst" and "regular" tokens.

gmarek replied:
I probably wrote something too complex. Rate limiter has to 'limits' one that is sustainable and 'burst' limit for short bursts. I think that exposing `number of tokens / max number of tokens` (something like 'burst saturation') is not enough. We want to know when RL allowed more QPS than it's sustainable level. There's no difference between tokens themselves, but there's a difference between having 1 QPS and 10 QPS with RL producing 2tok/sec and having burst value of 20. Does it make more sense now?

wojtek-t replied:
I initially thought I agree with @gmarek  but after more thinking about it, I think that I don't agree.

I think that the reason behind having this metric is to expose how many operations we are still able to process. And we are not going to look into a single value of the metric, but its graph over time.

So if the saturation is going up, it's a clear signal that there are more requests that we can actually handle (which is obviously a bad sign).

I think that we should leave as is (and potentially add another metric (something similar to what @gmarek suggested) in the future).

lavalamp replied:
@gmarek No, it still doesn't make sense to me. All that matters is the aggregate qps. If the actions taken are bursty then that will show up as sometimes being saturated and sometimes not. Since we poll this every second and we will look at an aggregate value, I don't think we need to add any logic to consider that case to this saturation function; it should return instantaneous saturation, as it does.

IOW, bursting really doesn't change the logic because you can't sustain it-- either you'll run out of tokens eventually and it will show up in this metric OR you're not actually exceeding the QPS limit and therefore we don't care.

hongchaodeng commented:
Thanks @timothysc @lavalamp @gmarek @wojtek-t for your comments.
Addressed them.

lavalamp commented on plugin/pkg/scheduler/metrics/metrics.go:
```diff
@@ -68,3 +79,12 @@ func Register() {
+// Metrics is collected every "interval".
+func ReportRateLimiterSaturation(rl util.RateLimiter, interval time.Duration) {
+	for {
+		BindingRateLimiterSaturation.Set(rl.Saturation())
```
I wasn't clear enough before, sorry. This is what I was trying to ask for:

```go
func ReportRateLimiterSaturation(metricToUpdate metrictype, rl util.RateLimiter) {
  metricToUpdate.Set(rl.Saturation())
}

...

go util.Forever(func() {
  ReportRateLimiterSaturation(BindingRateLimiterSaturation, rl)
}, time.Second)
```

hongchaodeng replied:
This actually makes more sense.

hongchaodeng replied:
I guess we didn't really need to add another function but set metrics inside `util.Forever`.

hongchaodeng commented on plugin/pkg/scheduler/scheduler.go:
```diff
@@ -107,6 +107,9 @@ func New(c *Config) *Scheduler {
 // Run begins watching and scheduling. It starts a goroutine and returns immediately.
 func (s *Scheduler) Run() {
+	go util.Forever(func() {
+		metrics.BindingRateLimiterSaturation.Set(s.config.BindPodsRateLimiter)
```
@lavalamp 
I didn't know well enough all those utility functions. Thanks for suggesting it. It actually looks clearer simpler in this way.

lavalamp commented:
LGTM, thanks for bearing with me :)

lavalamp commented:
I'll let @wojtek-t make a final pass if he wishes :)

hongchaodeng commented:
@lavalamp Thank you too. I learned many things from reading feedback and comments :)

wojtek-t commented on pkg/util/throttle.go:
```diff
@@ -48,22 +53,36 @@ func NewFakeRateLimiter() RateLimiter {
+	avail := t.limiter.Available()
+	// handles token burst, which means not saturated at all.
+	if avail > capacity {
+		avail = capacity
```
hmm - to be honest, I think that what we should expose is not available/capacity, but available/(burst capacity)

WDYT?

lavalamp replied:
I believe this *is* available / burst capacity, no?

hongchaodeng replied:
You are both right. 
I just wrote it too fast..

wojtek-t replied:
SGTM - thx

wojtek-t commented:
@hongchaodeng - thanks a lot for fixes - it is much better now

I added some comments, but those are more philosophical ones. [If you don't agree with them, I'm fine with @lavalamp lgtm-ing this PR].

lavalamp commented:
LGTM

hongchaodeng committed:
Scheduler metrics: binding rate limiter saturation

wojtek-t commented:
Thanks - LGTM

wojtek-t commented:
@k8s-bot ok to test

gmarek commented:
Test failure looks like a real thing:
```
I1203 08:30:12.853367   22739 factory.go:173] creating scheduler with fit predicates 'map[NoDiskConflict:{} MatchNodeSelector:{} HostName:{} PodFitsHostPorts:{} PodFitsResources:{}]' and priority functions 'map[LeastRequestedPriority:{} BalancedResourceAllocation:{} SelectorSpreadPriority:{}]
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x745de1]

goroutine 55 [running]:
k8s.io/kubernetes/plugin/pkg/scheduler.func·002()
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/scheduler.go:111 +0x41
k8s.io/kubernetes/pkg/util.func·006()
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:126 +0x61
k8s.io/kubernetes/pkg/util.Until(0xc2088fc1a0, 0x3b9aca00, 0xc20804c480)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:127 +0x8f
k8s.io/kubernetes/pkg/util.Forever(0xc2088fc1a0, 0x3b9aca00)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:109 +0x3e
created by k8s.io/kubernetes/plugin/pkg/scheduler.(*Scheduler).Run
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/scheduler.go:113 +0xa7
```

hongchaodeng commented:
How to reproduce that?

gmarek commented:
It happened during the integration test, so `hack/test-integration.sh` should do the trick.

hongchaodeng commented:
ratelimiter could be nil. See: https://github.com/kubernetes/kubernetes/blob/41ebe6f1a3de295f56d18744cb241c1af5ec6656/test/integration/extender_test.go#L241

We should add a caveat comment.

hongchaodeng commented:
Fixed
@gmarek

wojtek-t commented:
@hongchaodeng - thanks!

LGTM

gmarek commented:
@k8s-bot test this
